pkg/syncer/metadatasyncer.go

```diff
@@ -2640,3 +2655,139 @@ func initResizeReconciler(ctx context.Context, tkgClient clientset.Interface,
+	extensionName string) (*storagepolicyusagev1alpha1.StoragePolicyUsage, error) {
+	newUsageInstance := storagepolicyusagev1alpha1.StoragePolicyUsage{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       cnsoperatorv1alpha1.GroupName,
+			APIVersion: cnsoperatorv1alpha1.Version,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              name,
+			Namespace:         namespace,
+			Generation:        0,
+			CreationTimestamp: metav1.Time{},
+		},
+		Spec: storagepolicyusagev1alpha1.StoragePolicyUsageSpec{
+			StoragePolicyId:       storagePolicyId,
+			StorageClassName:      storageClassName,
+			ResourceKind:          resourceKind,
+			ResourceAPIgroup:      &resourceApiGroup,
+			ResourceExtensionName: extensionName,
+		},
+	}
+	err := quotaClient.Create(ctx, &newUsageInstance, &client.CreateOptions{})
+	if err != nil {
+		log.Errorf("Failed to create StoragePolicyUsage for policyID %v storageclass %v resourceKind %v. Err: %+v",
+			storagePolicyId, storageClassName, resourceKind, err)
+		return nil, err
+	}
+	return &newUsageInstance, nil
+}
+
+// policyQuotaCRAdded, on creation of StoragePolicyQuota object, creates StoragePolicyUsage object
+// for the Storage Policy Id and storage classes asspociated, in given same namespace as of StoragePolicyQuota object.
+func policyQuotaCRAdded(obj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Verify object received.
+	var policyQuotaObj storagepolicyusagev1alpha1.StoragePolicyQuota
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object,
+		&policyQuotaObj)
+	if err != nil {
+		log.Errorf("policyQuotaCRAdded: failed to cast object %+v to %s. Error: %v", obj,
+			cnsoperatorv1alpha1.CnsStoragePolicyQuotaSingular, err)
+		return
+	}
+	namespace := policyQuotaObj.Namespace
+	storagePolicyId := policyQuotaObj.Spec.StoragePolicyId
+	// Get storage classes associated with storage policy id of StoragePolicyQuota CR added
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("policyQuotaCRAdded: Failed to get KubeConfig. err: %v", err)
+		return
+	}
+	k8sClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		log.Errorf("policyQuotaCRAdded: Failed to create kubernetes client. Err: %+v", err)
+		return
+	}
+	storageQuotaClient, err := k8s.NewClientForGroup(ctx, config, cnsoperatorv1alpha1.GroupName)
+	if err != nil {
+		log.Errorf("policyQuotaCRAdded: Failed to create CnsOperator client. Err: %+v", err)
+		return
+	}
+	storageClassList, err := k8sClient.StorageV1().StorageClasses().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("policyQuotaCRAdded: Failed to list storageclasses. Err: %+v", err)
+		return
+	}
+	// For each storage class associated with storage policy id of StoragePolicyQuota CR,
+	// check if StoragePolicyUsage CR with resource type PVC exists.
+	// If not, create one with all parameters specified.
+	for _, sc := range storageClassList.Items {
```

Comment:
why do we need this loop to check all storageclasses when we are getting informer call upon creation of new `policyQuotaCRAdded`?

Is it because when driver restarts add event is replayed and you want to ignore re-creating usage CR instance again?

Reply:
When a new StoragePolicyQuota is created, we need to find the matching StorageClasses based on policyID, and create StoragePolicyUsage CRs and that's the reason for looping through the storageclass.
Are you suggesting that the storageclassname comes with the policyQuotaCR that got added ?

@xing-yang Can you confirm this ?

Reply:
Looping thru is a slower but safer way to identify the StorageClasses created by WCP based on policyID. In the future, maybe we can simplify this by looking up the storageClassName itself once we've confirmed the StorageClass naming convention.